Release 1.32.0 doc changes (#102)
nginx/unit-docs: source/howto/samples.rst

WebAssembly
***********

.. tabs::
   :prefix: web-assembly
   :toc:

   .. tab:: wasm-was-component

      Instead of dealing with bytecode,
      let's build a Unit-capable Rust app
      and compile it into WebAssembly Component.

      Make sure you have general Rust-Tooling (cargo, rustc, …) installed.

>       We recommend to use `rustup <https://rustup.rs/>`__ to get started.

Reviewer: ```suggestion
      We recommend using `rustup <https://rustup.rs/>`__ to get started.
```

      This example was build with **rustc** version 1.76.0.

      First, install the WebAssembly-specific Rust tooling:

      Start by adding the wasm32-wasi support as a compilation target for **rustc**

      .. code-block:: console

         $ rustup target add wasm32-wasi

>       Next, install **cargo component**. This will make it easier to build a

Reviewer: Suggested rewrite:  Next, install **cargo component**. This simplifies building a Wasm Component from Rust Code, making it the recommended method.

>       Next, install **cargo component**. This will make it easier to build a
>       Wasm Component from the Rust Code and is the preferred way of doing it.

Author: ```suggestion
      Next, install cargo component. This simplifies building a Wasm
      Component from Rust Code, making it the recommended method.
```

      .. code-block:: console

         $ cargo install cargo-component

>       Currently, the fastes way to get started with Wasm Components using WASI

Reviewer: ```suggestion
      Currently, the fastest way to get started with Wasm Components using WASI
```

      0.2 wasi-http API is the **hello-wasi-http** demo application by
      sunfishcode. Simply clone the repository and build the component running
      the following command:

      .. code-block:: console

         $ git clone https://github.com/sunfishcode/hello-wasi-http
         $ cd hello-wasi-http
         $ cargo component build

      The output of the build command should be similar to this:

      .. code-block:: console

         $ cargo component build
         Compiling hello-wasi-http v0.0.0 (/home/unit-build/hello-wasi-http)
         Finished dev [unoptimized + debuginfo] target(s) in 0.17s
         Creating component /home/unit-build/hello-wasi-http/target/wasm32-wasi/debug/hello_wasi_http.wasm
         $

>       This will create a WebAssembly Component you can deploy on Unit using the

Reviewer: ```suggestion
      This creates a WebAssembly Component you can deploy on Unit using the
```

Reviewer: Use present tense predominately.

      following Unit configuration. Make sure you point the **component** path
      to the WebAssembly component you have just created. Create a
      **config.json** file:

      .. code-block:: json

         {
            "listeners": {
               "127.0.0.1:8080": {
                  "pass": "applications/wasm"
               }
            },
            "applications": {
               "wasm": {
                  "type": "wasm-wasi-component",
                  "component": "/home/unit-build/hello-wasi-http/target/wasm32-wasi/debug/hello_wasi_http.wasm"
               }
            }
         }

      Apply the Unit configuration by using the CLI:

      .. code-block:: console

         $ unitc /config < config.json

      Or by sending it manually to Units control API:

      .. code-block:: console

         $ cat config.json | curl -X PUT -d @- --unix-socket /path/to/control.unit.sock http://localhost/config/

      Congratulations! You have created your very first WebAssembly Component
      on Unit! Send a GET Request to your configured listener.

      .. code-block:: console

         $ curl http://localhost:8080

   .. tab:: unit-wasm

      .. warning::

>          Unit 1.32 and later support the WebAssembly Component

Reviewer 2: `1.32.0`

Author: ```suggestion
         Unit 1.32.0 and later support the WebAssembly Component
```

      This yields the
      **target/wasm32-wasi/debug/wasm_on_unit.wasm** file
      (path may depend on other options).

      Upload the :ref:`app config <configuration-wasm>` to Unit and test it:

      .. code-block:: console

         # curl -X PUT --data-binary '{
               "listeners": {
                  "127.0.0.1:8080": {
                     "pass": "applications/wasm"
                  }
               },

               "applications": {
                  "wasm": {
                     "type": "wasm",
                     "module": ":nxt_ph:`/path/to/wasm_on_unit <app directory>`/target/wasm32-wasi/debug/wasm_on_unit.wasm",
                     "request_handler": "uwr_request_handler",
                     "malloc_handler": "luw_malloc_handler",
                     "free_handler": "luw_free_handler",
                     "module_init_handler": "uwr_module_init_handler",
                     "module_end_handler": "uwr_module_end_handler"
                  }
               }
         }' --unix-socket :nxt_ph:`/path/to/control.unit.sock <Path to Unit's control socket in your installation>` http://localhost/config/

         $ curl http://localhost:8080

>                * Welcome to WebAssembly in Rust on Unit! [libunit-wasm (0.1.0/0x00010000)] *

Reviewer 2: Perhaps change the version to `(0.3.0/0x00030000)` which is the latest...

Author: ```suggestion
               * Welcome to WebAssembly in Rust on Unit! [libunit-wasm (0.3.0/0x00030000)] *
```

               [Request Info]
               REQUEST_PATH = /
               METHOD       = GET
               VERSION      = HTTP/1.1
               QUERY        =
               REMOTE       = 127.0.0.1
               LOCAL_ADDR   = 127.0.0.1
               LOCAL_PORT   = 8080

>                SERVER_NAME  = localhost

Reviewer 2: The Full response would be like

```
[Request Info]
REQUEST_PATH = /
METHOD       = GET
VERSION      = HTTP/1.1
QUERY        = 
REMOTE       = 127.0.0.1
LOCAL_ADDR   = 127.0.0.1
LOCAL_PORT   = 8080
SERVER_NAME  = localhost

[Request Headers]
Host = localhost:8080
User-Agent = curl/8.2.1
Accept = */*
```

Author: ```suggestion

               [Request Info]
               REQUEST_PATH = /
               METHOD       = GET
               VERSION      = HTTP/1.1
               QUERY        =
               REMOTE       = 127.0.0.1
               LOCAL_ADDR   = 127.0.0.1
               LOCAL_PORT   = 8080
               SERVER_NAME  = localhost
               
               [Request Headers]
               Host = localhost:8080
               User-Agent = curl/8.2.1
               Accept = */*
```

WebAssembly
***********

.. tabs::
   :prefix: web-assembly
   :toc:

>    .. tab:: wasm-was-component

Reviewer 2: ```suggestion
   .. tab:: wasm-wasi-component
```

      Instead of dealing with bytecode,
      let's build a Unit-capable Rust app

>       and compile it into a WebAssembly Component.

Reviewer 3: ```suggestion
      and compile it into a WebAssembly (Wasm) Component.
```

Author: ```suggestion
      and compile it into a WebAssembly (Wasm) component.
```

>       Make sure you have general Rust-Tooling (cargo, rustc, …) installed.

Reviewer 3: ```suggestion
      Make sure you have the Rust toolchain (cargo, rustc, …) installed.
```
Unless `Rust-Tooling` is a known thing (with the hyphen).

Reviewer 4: LGTM

Author: ```suggestion
      Make sure you have the Rust toolchain (cargo, rustc, etc.) installed.
```
Taking your suggestion but removing the ellipsis which should be avoided

      We recommend using `rustup <https://rustup.rs/>`__ to get started.

      This example was built with **rustc** version 1.76.0.

>       First, install the WebAssembly-specific Rust tooling:

Reviewer 3: Huh, didn't we just do that on line 565? And why the colon?

Author: removed

>       First, install the WebAssembly-specific Rust tooling:
> 

Author: ```suggestion

```
remove the line as per previous suggestion

      Start by adding the wasm32-wasi support as a compilation target for **rustc**

      .. code-block:: console

         $ rustup target add wasm32-wasi

>       Next, install cargo component. This simplifies building a Wasm

Reviewer 3: This is the first mention of "Wasm" but we didn't explain that it's an abbreviation of WebAssembly.

>       Next, install cargo component. This simplifies building a Wasm

Author: we are adding WebAssembly (Wasm) in line 563 as per your suggestion.
Do you think we need to add it here too?

Reviewer 3: No. I think we should take one of these approaches:

1. Use WebAssembly everywhere and never mention "Wasm".
2. Introduce the abbreviation "WebAssembly (Wasm)" and then use "Wasm" from that point on (never mentioning "WebAssembly" again).

Author: I checked the WebAssembly documentation. They start with WebAssembly (Wasm) and then stick to WebAssembly through the doc.

I have updated the doc to follow the same pattern. Resolving this.

      Component from Rust Code, making it the recommended method.

      .. code-block:: console

         $ cargo install cargo-component

      Currently, the fastest way to get started with Wasm Components using WASI
      0.2 wasi-http API is the **hello-wasi-http** demo application by

>       sunfishcode. Simply clone the repository and build the component running

Reviewer 5: ```suggestion
      Dan Gohman. Simply clone the repository and build the component running
```

...maybe?

Reviewer 5: Also, do we want to avoid language like "simply"?

Author: ```suggestion
      Dan Gohman. Clone the repository and build the component running
```

      the following command:

      .. code-block:: console

         $ git clone https://github.com/sunfishcode/hello-wasi-http
         $ cd hello-wasi-http
         $ cargo component build

      The output of the build command should be similar to this:

      .. code-block:: console

         $ cargo component build
         Compiling hello-wasi-http v0.0.0 (/home/unit-build/hello-wasi-http)
         Finished dev [unoptimized + debuginfo] target(s) in 0.17s
         Creating component /home/unit-build/hello-wasi-http/target/wasm32-wasi/debug/hello_wasi_http.wasm
         $

>       This creates a WebAssembly Component you can deploy on Unit using the

Reviewer 3: Why are we flip-flopping between Wasm and WebAssembly?

Reviewer 4: No reason. But you are right, having a coherent wording would be better! @Jcahilltorre i am good using WebAssembly.

Author: I will update this via a commit to use WebAssembly only (same as the WebAssembly docs) *(except when referring to module names, processes, etc)
